### ISSUE TITLE
static-location replacement isn't well-suited to testing

### ISSUE THREAD

**Issue by davidje13** [molefrog/wouter#330]
in #113 I suggested (then implemented) a location history for `static-location`, but this helper has been replaced with `ssrPath` in the latest version (and marked deprecated). `ssrPath` doesn't provide the same testing abilities (i.e. detecting URL changes), so I'd like to propose un-deprecating that helper for tests (maybe with a rename so that people are still forced to consider using the new API for SSR)

Alternatively, it could be useful to provide the ability to detect URL changes when using `ssrPath` (as noted in the original issue: the non-test use-case would be detecting a page which redirects to another, and sending a 3xx response instead of the content)

(in my tests I have checks which fail if any console messages are printed, so the deprecation note makes them fail by itself!)

**Comment by molefrog** [comment]
Hi @davidje13, sorry that we had to remove your implementation, the problem was that `static-location` didn't really make sense with the new release, since a lot of users were using it to achieve SSR which now should be done by specifying `ssrPath`.

I definitely agree we should get the testing helpers back. Do you have ideas regarding the new API?

**Comment by davidje13** [comment]
I think the cleanest API to match the new `ssrPath` property would be to add a callback, like:

```jsx
<Router ssrPath="/somewhere" onLocationChange={myCallback}>...
```

Then in testing, a user could write code like:

```jsx
const locationChangeFn = jest.fn();
render(<Router ssrPath="/somewhere" onLocationChange={locationChangeFn}>...</Router>);

// do stuff

expect(locationChangeFn).toHaveBeenCalledTimes(1);
expect(locationChangeFn).toHaveBeenCalledWith('/my-new-location');
```

For SSR redirects, it would look something like:

```jsx
let redirect = null;
const rendered = renderToString(<Router ssrPath="/somewhere" onLocationChange={(location) => redirect = location}>...</Router>);
if (redirect) {
  // return 3xx
} else {
  // return content
}
```

And I could imagine the same handler being used for application stuff (e.g. analytics tracking).

I guess ideally the same callback would be invoked no matter how the location was changed (programmatically, or from the URL changing, or from navigating history), but I could imagine future use-cases that might want to distinguish between these cases so it might be good to have a second argument in the callback which says what sort of operation caused the change.

**Comment by molefrog** [comment]
I've just published the [v3.0.0-rc.1](https://github.com/molefrog/wouter/releases/tag/v3.0.0-rc.1) that:

- adds SSR support for search strings (`ssrSearch`)
- removes static location
- introduces `"wouter/memory-location"` that can be used in testing, as a static location or in-memory location with `recor` option

See https://github.com/molefrog/wouter/issues/355 for more details
